tiny_cnn/core/tiny_backend.h

```diff
@@ -0,0 +1,150 @@
+*/
+#pragma once
+
+#include "tiny_cnn/core/math_backend.h"
+
+namespace tiny_cnn {
+namespace core {
+
+class tiny_backend : public math_backend {
+ public:
+  // context holds solution-dependent parameters
+  // context should be able to hold any types of structures (like boost::any)
+  tiny_backend(const conv_params& params) : params_(params) {}
+
+  // core math functions
+
+  void conv2d(cnn_size_t                 index,
+              const std::vector<vec_t*>& in_data,
+              std::vector<vec_t*>&       out_data) {
+      copy_and_pad_input(*in_data[0], static_cast<int>(index));
+      const vec_t& W   = *in_data[1];
+      //vec_t&       out = *out_data[0];
+      vec_t&       a   = *out_data[1];
+      const vec_t &in  = *(params_.conv_layer_worker_storage[index].prev_out_padded_); // input //NOLINT
+
+      cnn_size_t idx = 0;
+      std::fill(a.begin(), a.end(), float_t(0));
+
+      for_i(layer_->get_parallelize(), params_.out.depth_, [&](int o) {
+          for (cnn_size_t inc = 0; inc < params_.in.depth_; inc++) {
+              if (!params_.tbl.is_connected(o, inc)) continue;
+
+              idx = params_.in.depth_ * o + inc;
+              idx = params_.weight.get_index(0, 0, idx);
+              const float_t *pw = &W[idx];
+
+              idx = params_.in_padded.get_index(0, 0, inc);
+              const float_t *pi = &in[idx];
+
+              idx = params_.out.get_index(0, 0, o);
+              float_t *pa = &a[idx];
+
+              for (cnn_size_t y = 0; y < params_.out.height_; y++) {
+                  for (cnn_size_t x = 0; x < params_.out.width_; x++) {
+                      const float_t * ppw = pw;
+                      const float_t * ppi = pi + params_.in_padded.width_ *
+                                            (y * params_.h_stride) +
+                                             x * params_.w_stride;
+                      float_t sum = float_t(0);
+
+                      // should be optimized for small kernel(3x3,5x5)
+                      for (cnn_size_t wy = 0; wy < params_.weight.height_; wy++) {    //NOLINT
+                          for (cnn_size_t wx = 0; wx < params_.weight.width_; wx++) { //NOLINT
+                              idx = wy * params_.in_padded.width_ + wx;
+                              sum += *ppw++ * ppi[idx];
+                          }
+                      }
+                      pa[y * params_.out.width_ + x] += sum;
+                  }
+              }
+          }
+
+          if (params_.has_bias) {
+              const vec_t& bias = *in_data[2];
+              float_t * pa  = &a[params_.out.get_index(0, 0, o)];
+              float_t * paa = pa + params_.out.width_ * params_.out.height_;
+              std::for_each(pa, paa, [&](float_t& f) { f += bias[o]; });
+          }
+      });
+
+      /*for_i(this->parallelize_, params.out.size(), [&](int i) {
+          out[i] = this->h_.f(a, i);
+      });*/
+  }
```

Comment:
@nyanp not sure if it should be done here or outside the kernel. Besides, check how do I use conv params. Also not sure if they should be in kernel signature.